### ISSUE TITLE
Write EOF error when comitting

### ISSUE THREAD

**Issue by cloewen8** [desktop/desktop#2041]
### Description

I get the following error when making a commit. The error occurs consistently.

> Error: write EOF
>     at exports._errnoException (util.js:1022:11)
>     at WriteWrap.afterWrite (net.js:804:14)
> 
> Version: 0.6.1
> OS: Windows 10.0.15063

The error may be related to the amount of staged files (3846 files). I am making the commit to remove the unused template. Some files may be missing from the template.

### Version

**OS version:** Microsoft Windows [Version 10.0.15063]

### Steps to Reproduce

The template was setup by OpenShift using the repository at [https://github.com/strongloop/openshift-cartridge-strongloop](https://github.com/strongloop/openshift-cartridge-strongloop).

OpenShift creates a new git repository using the manifest.yml file (located at [https://raw.github.com/strongloop/openshift-cartridge-strongloop/master/metadata/manifest.yml](https://raw.github.com/strongloop/openshift-cartridge-strongloop/master/metadata/manifest.yml).

The same repository structure can be achieved by:

1. Creating a new local repository.
2. Copying the files from [https://github.com/strongloop/openshift-cartridge-strongloop/tree/master/template](https://github.com/strongloop/openshift-cartridge-strongloop/tree/master/template).
3. Staging and committing the added files.

Attempting to commit the removal of all of the files causes the error to occur.

**Expected behaviour:** GitHub for Windows should show a waiting status, then a few seconds later the files should be successfully committed.

**Actual behaviour:** GitHub for Windows closes after waiting a few seconds with the same error message.

**Reproduces how often:** This error occurs consistently.

### Logs

The error logged stays consistent:

> 2017-06-18T04:58:02.738Z - info: [ui] Stats reported.
> 2017-06-18T04:58:47.905Z - error: [main] Error: write EOF
>     at exports._errnoException (util.js:1022:11)
>     at WriteWrap.afterWrite (net.js:804:14)
> 2017-06-18T04:58:47.940Z - info: [main] Error report submitted

#### Additional Information

A package.json and README.md file were added in the same commit (unchecking them yields the same result).

**Comment by cjcool** [comment]
I tried adding the files to a commit manually. Git says, 

> Filename too long

**Comment by shiftkey** [comment]
@cjcool thanks for the detailed report! this seems like enough to reproduce the issue but I'll report back if I can't see the same behaviour.

**Comment by joshaber** [comment]
This is one of our most common exceptions, so it's great to get some reproduction steps. Thanks @cjcool!

**Comment by shiftkey** [comment]
This is an interesting example repository, as trying to clone this to a very short folder like `C:/c/l` raises `MAX_PATH` issues on Windows:

```sh
$ $ git clone  https://github.com/strongloop/openshift-cartridge-strongloop.git l
remote: Counting objects: 22615, done.
remote: Total 22615 (delta 0), reused 0 (delta 0), pack-reused 22615
Receiving objects: 100% (22615/22615), 88.15 MiB | 1.73 MiB/s, done.
Resolving deltas: 100% (5940/5940), done.
fatal: cannot create directory at 'strongloop/node_modules/generator-loopback/node_modules/loopback-workspace/node_modules/loopback-boot/node_modules/lodash.clonedeep/node_modules/lodash._baseclone/node_modules/lodash.assign/node_modules/lodash.keys/node_modules/lodash._isnative': Filename too long
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

**Comment by shiftkey** [comment]
@cjcool I'm having some issues recreating the issue under development mode, and I think it's related to the `MAX_PATH` issue as reported above.

Here's my setup:

 - clone the repository in macOS, where we don't encounter `MAX_PATH` constraints
 - open my Win10 development VM, checkout `release-0.6.1` and `npm install && npm run build:dev && npm start` to get it running
 - initialize an empty repository at `C:/c/a` to maximize the remaining path characters available
 - copy the `template` folder over from the macOS filesystem (mounted as a shared folder in the VM) - it warns that some paths are too long, which is to be expected, and so I don't have much choice here aside from skipping

<img width="474"  src="https://user-images.githubusercontent.com/359239/27494421-a5981834-5802-11e7-973c-dc67b5f9e919.png">

 - After this completes, launch Desktop and drag-and-drop the folder. We've now got 3877 files to commit, which seems enough to show the issue. I'm going to stage and commit them all.

<img width="311" alt="screen shot 2017-06-23 at 10 53 16 am" src="https://user-images.githubusercontent.com/359239/27494452-ce2b953c-5802-11e7-9b47-07db0a1f025b.png">

 - After setting a commit summary, we can start the commit and it actually completes, and I don't see your error:

<img width="1108" src="https://user-images.githubusercontent.com/359239/27494479-e9d3d2f4-5802-11e7-87f0-a58169903107.png">

So my current question is: **is there a different way that you're creating the directory structure that works around `MAX_PATH` limit that Windows is complaining about?**

**Comment by cjcool** [comment]
I have the repository created by OpenShift (using the manifest.yml file mentioned in the original comment). It appears OpenShift uses a linux based operating system (https://blog.openshift.com/the-os-behind-the-curtain/).

**Comment by shiftkey** [comment]
> It appears OpenShift uses a linux based operating system (https://blog.openshift.com/the-os-behind-the-curtain/).

This still doesn't quite explain how to get the repository onto the Windows filesystem, and I couldn't trigger this error using a similar number of new files like the repro steps suggested. Any other clues would be greatly appreciated!

**Comment by cjcool** [comment]
I got the files into the Windows file system using git clone (using a provided ssh url). If you need any more specific information, I could try and gather it.

**Comment by scysys** [comment]
Same here

```
Error: write EOF
    at exports._errnoException (util.js:1022:11)
    at WriteWrap.afterWrite (net.js:804:14)

Version: 0.7.0
OS: Windows 10.0.15063
```
**Happens when Filename was too long**
error: open("app/build/tmp/expandedArchives/classes.jar_2piwhg11dmgkuwsgpw9scic0s/com/google/protobuf/java_com_google_android_gmscore_dev_policy_sets__3p_monolithic_raw_pre_munge_rollup_67e347e9GeneratedExtensionRegistryLite$Loader.class"): Filename too long

**Comment by shiftkey** [comment]
The more I think about this issue, the less things make sense:

```
Error: write EOF
at exports._errnoException (util.js:1022:11)
at WriteWrap.afterWrite (net.js:804:14)
```

`net.js` has nothing to do with the filesystem - and the problem code is [here](https://github.com/nodejs/node/blob/v7.4.0/lib/net.js#L803-L808). A rudimentary Google shows few clues as to the root cause, like https://github.com/imagemin/imagemin-jpegoptim/issues/7 (that library is also working with the filesystem) so I'm going to ignore that for now.

There's two things I can think of which could be to blame here (and it may be one or both or none of these):

 - we don't enable `core.longpaths` for Windows, which relaxes the `MAX_PATH` restriction for Git but is also Danger Zone™. These are the docs:

<img width="955" src="https://user-images.githubusercontent.com/359239/28534279-93a8f756-7076-11e7-8c8c-a74de669d8b9.png">
 
@scysys could you enable this config value on your problem repository and see if it changes the behaviour for that file path? Also, I'm curious about how you were able to create the path for that file to bypass `MAX_PATH` - what tooling or IDE are you using there? Can you recreate it with a public repository?

 - the other common thread seems to be related to the number of files used in a commit. #2286 mentions 2000+ files, but I wasn't able to reproduce this with 3877 files earlier so I'm less inclined to go down that path again. But keep in mind that we're using `stdin` to pass in all these [file paths](https://github.com/desktop/desktop/blob/master/app/src/lib/git/update-index.ts#L92-L94) before committing - and [`child.stdin`](https://nodejs.org/api/child_process.html#child_process_child_stdin) is a writeable stream, which kind of fits with the error message (if you ignore the stack trace).

**Comment by scysys** [comment]
> @scysys could you enable this config value on your problem repository and see if it changes the behaviour for that file path?

You right when I am adding "longpaths = true" to repro config, it was working without problems.

> Also, I'm curious about how you were able to create the path for that file to bypass MAX_PATH - what tooling or IDE are you using there?

It was Android Studio in his actual version. I checked it again now and deleted the old "tmp" folder. Android Studio seems to be ignoring max_path values under Windows.

**Android Studio**
`D:\Cloud\OneDrive - xxxxxxxxxxx\Dev\Github\Android-Customer-App\app\build\tmp\expandedArchives\classes.jar_2piwhg11dmgkuwsgpw9scic0s\com\google\protobuf\java_com_google_android_gmscore_dev_policy_sets__3p_monolithic_raw_pre_munge_rollup_67e347e9GeneratedExtensionRegistryLite$Loader.class`

**When I try to create this file myself**
`D:\Cloud\OneDrive - xxxxxxxxxxx\Dev\Github\Android-Customer-App\app\build\tmp\expandedArchives\classes.jar_2piwhg11dmgkuwsgpw9scic0s\com\google\protobuf\ava_com_google_android_gmscore_dev_policy_sets__3p_monolithic_raw_pre_munge_rollup_67e347e9Gen`

> the other common thread seems to be related to the number of files used in a commit. #2286 mentions 2000+ files, but I wasn't able to reproduce this with 3877 files earlier so I'm less inclined to go down that path again.

It has nothing to do with number of files. The commit with this problem has over 5000 files. Commit was working without problems when I remove the files, where was filename too long.

**Comment by shiftkey** [comment]
@scysys thanks for the insight - this should help me get to the bottom of reproducing the issue and proposing a way out!

**Comment by shiftkey** [comment]
Made some progress on this, however it's not the original `error: write EOF` I was hoping for.

Here's the problem file path (full path is 353 characters, well over `MAX_PATH`):

<img width="1047" src="https://user-images.githubusercontent.com/359239/28572802-e2187ddc-711e-11e7-847c-0eb3e6948453.png">

When trying to commit it, you get the Git output in the error dialog:

<img width="616" src="https://user-images.githubusercontent.com/359239/28572856-0f05b92c-711f-11e7-976b-7a0633ff39d7.png">

And the log file has the same info:

```
2017-07-25T12:50:10.205Z - error: [ui] `git update-index --add --remove --replace -z --stdin` exited with an unexpected code: 128.
warning: LF will be replaced by CRLF in .idea/compiler.xml.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in .idea/gradle.xml.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in .idea/misc.xml.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in .idea/modules.xml.
The file will have its original line endings in your working directory.
warning: LF will be replaced by CRLF in .idea/runConfigurations.xml.
The file will have its original line endings in your working directory.
error: lstat("app/src/main/java/com/example/shiftkey/myapplication/HelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorld.java"): Filename too long
fatal: Unable to process path app/src/main/java/com/example/shiftkey/myapplication/HelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorldHelloWorld.java
```

And here's the underlying CLI experience (with the workaround that @scysys confirmed earlier):

<img width="589" src="https://user-images.githubusercontent.com/359239/28572965-5990b0be-711f-11e7-8d4c-6b75cc337f52.png">

I'll go down the path of addressing this issue, but the fact that it's not the `error: write EOF` error saddens me greatly.

**Comment by beniroquai** [comment]
Same here:

Error: write EOF
    at exports._errnoException (util.js:1050:11)
    at WriteWrap.afterWrite (net.js:813:14)

Version: 1.0.8
OS: Windows 10.0.15063


I'm syncing stuff with Dropbox from MAC to Windows and try to commit from Windows side. Could this be an issue?
May thanks!

**Comment by shiftkey** [comment]
@beniroquai can you please share what sort of files you are trying to commit? For example, #3370 talks a bunch about `node_modules` being problematic.

**Comment by DavidBurela** [comment]
I was getting this issue also, had discovered that my .gitignore file had been deleted and so it was trying to check in the node_modules folder. Ignoring that folder let me commit again without crashes

**Comment by shiftkey** [comment]
We made a fix to our embedded Git in https://github.com/desktop/dugite/pull/164 which seems to address the `write EOF` error along with some others, and this is available in `1.1.0` if you update your installs.

If you are still able to trigger this error, please open a new issue with details and we can try and get to the bottom of it.
